src/libraries/Microsoft.PowerFx.Core/IR/BinaryOpMatrix.cs

```diff
@@ -0,0 +1,393 @@
+                    {
+                        case BinaryOp.NotEqual:
+                            return BinaryOpKind.NeqMedia;
+                        case BinaryOp.Equal:
+                            return BinaryOpKind.EqMedia;
+                        default:
+                            throw new NotSupportedException();
+                    }
+
+                case DKind.Blob:
+                    switch (node.Op)
+                    {
+                        case BinaryOp.NotEqual:
+                            return BinaryOpKind.NeqBlob;
+                        case BinaryOp.Equal:
+                            return BinaryOpKind.EqBlob;
+                        default:
+                            throw new NotSupportedException();
+                    }
+
+                case DKind.Guid:
+                    switch (node.Op)
+                    {
+                        case BinaryOp.NotEqual:
+                            return BinaryOpKind.NeqGuid;
+                        case BinaryOp.Equal:
+                            return BinaryOpKind.EqGuid;
+                        default:
+                            throw new NotSupportedException();
+                    }
+
+                case DKind.ObjNull:
+                    switch (node.Op)
+                    {
+                        case BinaryOp.NotEqual:
+                            return BinaryOpKind.NeqNull;
+                        case BinaryOp.Equal:
+                            return BinaryOpKind.EqNull;
+                        default:
+                            throw new NotSupportedException();
+                    }
+
+                case DKind.OptionSetValue:
+                    switch (node.Op)
+                    {
+                        case BinaryOp.NotEqual:
+                            return BinaryOpKind.NeqOptionSetValue;
+                        case BinaryOp.Equal:
+                            return BinaryOpKind.EqOptionSetValue;
+                        default:
+                            throw new NotSupportedException();
+                    }
+
+                default:
+                    throw new NotSupportedException("Not supported comparison op on type " + kindToUse.ToString());
+            }
+        }
+
+        private static BinaryOpKind GetAddOp(Syntax.BinaryOpNode node, DType leftType, DType rightType)
+        {
+            switch (leftType.Kind)
+            {
+                case DKind.Date:
+                    if (rightType == DType.DateTime || rightType == DType.Date)
+                    {
+                        // Date + '-DateTime' => in days
+                        // Date + '-Date' => in days
+
+                        // Binding produces this as '-Date'. This should be cleaned up when we switch to a proper sub op. 
+                        if (node.Right is not Syntax.UnaryOpNode { Op: UnaryOp.Minus })
+                        {
+                            throw new NotSupportedException();
```

Comment:
Why are Exceptions being used in this method? The other methods don't throw exceptions, and instead return Invalid.

Reply:
Preserving existing behavior, but these scenarios are not representable from the parser/binder. Invalid is not valid, but still representable with errors from the parser/binder.